Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
         m_contentSecurityPolicy = "script-src 'self'"_s;
 }
 
+bool WebExtension::hasBackgroundContent()
+{
+    populateBackgroundPropertiesIfNeeded();
+    return m_backgroundScriptPaths.size() || !m_backgroundPagePath.isEmpty() || !m_backgroundServiceWorkerPath.isEmpty();
+}
+
+bool WebExtension::backgroundContentIsPersistent()
+{
+    populateBackgroundPropertiesIfNeeded();
+    return hasBackgroundContent() && m_backgroundContentIsPersistent;
+}
+
+bool WebExtension::backgroundContentUsesModules()
+{
+    populateBackgroundPropertiesIfNeeded();
+    return hasBackgroundContent() && m_backgroundContentUsesModules;
+}
+
+bool WebExtension::backgroundContentIsServiceWorker()
+{
+    populateBackgroundPropertiesIfNeeded();
+    return m_backgroundContentEnvironment == Environment::ServiceWorker;
+}
+
+const String& WebExtension::backgroundContentPath()
+{
+    populateBackgroundPropertiesIfNeeded();
+
+    if (!m_backgroundServiceWorkerPath.isEmpty())
+        return m_backgroundServiceWorkerPath;
+
+    if (m_backgroundScriptPaths.size()) {
+        if (backgroundContentIsServiceWorker()) {
+            static const NeverDestroyed<String> backgroundContentString = generatedBackgroundServiceWorkerFilename;
+            return backgroundContentString;
+        }
+
+        static const NeverDestroyed<String> backgroundContentString = generatedBackgroundPageFilename;
+        return backgroundContentString;
+    }
+
+    if (!m_backgroundPagePath.isEmpty())
+        return m_backgroundPagePath;
+
+    ASSERT_NOT_REACHED();
+    return nullString();
+}
+
+const String& WebExtension::generatedBackgroundContent()
+{
+    if (!m_generatedBackgroundContent.isEmpty())
+        return m_generatedBackgroundContent;
+
+    populateBackgroundPropertiesIfNeeded();
+
+    if (!m_backgroundServiceWorkerPath.isEmpty() || !m_backgroundPagePath.isEmpty())
+        return nullString();
+
+    if (m_backgroundScriptPaths.size())
+        return nullString();
+
+    bool isServiceWorker = backgroundContentIsServiceWorker();
+    bool usesModules = backgroundContentUsesModules();
+
+    RefPtr scriptsArray = JSON::Array::create();
```

Comment:
This should just be a `Vector<String>`.
```suggestion
    Vector<String> scripts;
```